CRM/MembershipExtras/Hook/PostProcess/MembershipOfflineAutoRenewProcessor.php

```diff
@@ -0,0 +1,247 @@
+   * @var \CRM_Member_Form_Membership
+   *
+   * Form object that is being altered.
+   */
+  private $form;
+
+  /***
+   * @var array
+   *
+   * Form submitted data values.
+   */
+  private $formSubmittedValues;
+
+  public function __construct(&$form) {
+    $this->form = $form;
+    $this->formSubmittedValues = $this->form->exportValues();
+  }
+
+  /**
+   * Processes the membership new/renew form
+   * to handle offline auto-renewal.
+   */
+  public function process() {
+    if (!$this->isOfflineAutoRenewMembership() || $this->membershipIsAlreadyAutoRenew()) {
+      return;
+    }
+
+    $isPaymentPlanPayment = $this->isPaymentPlanPayment();
+    if ($isPaymentPlanPayment) {
+      $recurContributionID = $this->getMembershipLastRecurContributionID();
+    }
+    else {
+      $recurContributionID = $this->createAutoRenewRecurContribution();
+    }
+
+    $this->setMembershipToAutoRenew($recurContributionID);
+
+    if ($isPaymentPlanPayment) {
+      $this->setRecurContributionAutoRenew($recurContributionID);
+    }
+  }
+
+  /**
+   * Determines if the user selected
+   * offline auto-renew option or not.
+   *
+   * @return mixed
+   */
+  private function isOfflineAutoRenewMembership() {
+    return CRM_Utils_Request::retrieve('offline_auto_renew', 'Int');
+  }
+
+  /**
+   * Determines if the membership is
+   * already set to be auto renewed or not.
+   *
+   * @return mixed
+   */
+  private function membershipIsAlreadyAutoRenew() {
+    return CRM_Utils_Request::retrieve('membership_is_already_autorenew', 'Int');
+  }
+
+  /**
+   * Detects if the membership is paid for
+   * using payment plan option.
+   *
+   * @return bool
+   */
+  private function isPaymentPlanPayment() {
+    $installmentsCount = CRM_Utils_Request::retrieve('installments', 'Int');
+    $isSavingContribution = CRM_Utils_Request::retrieve('record_contribution', 'Int');
+    $contributionIsPaymentPlan = CRM_Utils_Request::retrieve('contribution_type_toggle', 'String') === 'payment_plan';
+
+    if ($isSavingContribution && $contributionIsPaymentPlan && $installmentsCount > 1) {
+      return TRUE;
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * Gets the membership last recurring contribution ID.
+   * (For Payment Plan payments)
+   *
+   * @return mixed
+   */
+  private function getMembershipLastRecurContributionID() {
+    $recurContributionID = civicrm_api3('MembershipPayment', 'get', [
+      'sequential' => 1,
+      'return' => ['contribution_id.contribution_recur_id'],
+      'membership_id' => $this->form->_id,
+      'options' => ['limit' => 1, 'sort' => 'contribution_id.contribution_recur_id DESC'],
+    ])['values'][0]['contribution_id.contribution_recur_id'];
+
+    return $recurContributionID;
+  }
+
+  /**
+   * Creates recurring contribution for the auto-renewed
+   * membership.
+   * (For contribution payment type)
+   *
+   * @return int
+   */
+  private function createAutoRenewRecurContribution() {
+    $PaymentInstrument = 'null';
```

Comment:
Variable names are supposed to start with lowercase letters by coding standards.